# Ejercicio 103

Generar una representación de los elementos de un `DataFrame` con `applymap`.

## Solución:

In [1]:
import pandas as pd

In [2]:
numeros = [[3.1415, 2.717], [1.41, 1]]

df = pd.DataFrame(numeros)

df

0      1
0  3.1415  2.717
1  1.4100  1.000

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 2 columns):
0    2 non-null float64
1    2 non-null float64
dtypes: float64(2)
memory usage: 112.0 bytes


In [4]:
df.applymap(lambda x: len(str(x)))

0  1
0  6  5
1  4  3

In [5]:
df.applymap(lambda x: x**2)

0         1
0  9.869022  7.382089
1  1.988100  1.000000

In [6]:
df ** 2

0         1
0  9.869022  7.382089
1  1.988100  1.000000

# Ejercicio 104

Usar `groupby` para agrupar datos por una columna específica y contar la cantidad de elementos.

## Solución:

In [7]:
import pandas as pd

In [8]:
lenguajes = ['Python', 'Python', 'Java', 'C#', 'Java', 'PHP', 'JavaScript', 'Python',
           'JavaScript', 'Java', 'C#', 'JavaScript']

programadores = ['Edward', 'Daniela', 'Angie', 'Germán', 'Valentina', 'Valeria',
                'Oliva', 'Diego', 'Andrés', 'Carlos', 'Mauricio', 'Guillermo']

datos = {'programador': programadores, 'lenguaje': lenguajes}

df = pd.DataFrame(data=datos)

df

programador    lenguaje
0       Edward      Python
1      Daniela      Python
2        Angie        Java
3       Germán          C#
4    Valentina        Java
5      Valeria         PHP
6        Oliva  JavaScript
7        Diego      Python
8       Andrés  JavaScript
9       Carlos        Java
10    Mauricio          C#
11   Guillermo  JavaScript

In [9]:
df.groupby(['lenguaje']).size().reset_index(name='numero_preferencias')

lenguaje  numero_preferencias
0          C#                    2
1        Java                    3
2  JavaScript                    3
3         PHP                    1
4      Python                    3

# Ejercicio 105

Usar la función `agg` para realizar varias operaciones por un eje de un `DataFrame`.

## Solución:

In [10]:
import pandas as pd
import numpy as np

In [11]:
datos = {'W': [1, 2, 3, 4, 5], 'X': [2, 3, 5, 7, 11],
         'Y': [5, 4, 3, 2, 1], 'Z': [np.nan, np.nan, np.nan, np.nan, np.nan]}

df = pd.DataFrame(data=datos)

df

W   X  Y   Z
0  1   2  5 NaN
1  2   3  4 NaN
2  3   5  3 NaN
3  4   7  2 NaN
4  5  11  1 NaN

In [12]:
df.agg(['sum', 'min', 'max'])

W   X   Y    Z
sum  15  28  15  0.0
min   1   2   1  NaN
max   5  11   5  NaN

In [13]:
df.aggregate(['sum', 'min', 'max'])

W   X   Y    Z
sum  15  28  15  0.0
min   1   2   1  NaN
max   5  11   5  NaN

In [14]:
df.agg({'W': ['sum'], 'X': ['sum', 'min', 'max'], 'Y': ['min', 'max'], 'Z': ['sum']})

W   X    Y    Z
max   NaN  11  5.0  NaN
min   NaN   2  1.0  NaN
sum  15.0  28  NaN  0.0

# Ejercicio 106

Eliminar registros que no cumplan con una condición sobre una columna específica.

## Solución:

In [15]:
import pandas as pd

In [16]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 5], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  5  11

In [17]:
df = df[df.Y != 5]

In [18]:
df

X  Y  Z
1  2  4  3
2  3  3  5
3  4  2  7

# Ejercicio 107

Transformar los valores de un `DataFrame` con la función `transform`.

## Solución:

In [19]:
import pandas as pd

In [20]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [21]:
df.transform(lambda x: x + 2)

X  Y   Z
0  3  7   4
1  4  6   5
2  5  5   7
3  6  4   9
4  7  3  13

In [22]:
def agregar_dos_unidades(x):
    return x + 2

df.transform(agregar_dos_unidades)

X  Y   Z
0  3  7   4
1  4  6   5
2  5  5   7
3  6  4   9
4  7  3  13

# Ejercicio 108

Agrupar un conjunto de datos y luego calcular su promedio con `groupby` y `mean`.

## Solución:

In [23]:
import pandas as pd

In [24]:
animales = ['Águila', 'Alcón', 'Águila', 'Alcón']
velocidades = [160, 390, 120, 385]

df = pd.DataFrame({'animal': animales, 'velocidad': velocidades})

df

animal  velocidad
0  Águila        160
1   Alcón        390
2  Águila        120
3   Alcón        385

In [25]:
df.groupby(['animal'])

In [26]:
df.groupby(['animal']).mean()

velocidad
animal           
Alcón       387.5
Águila      140.0

# Ejercicio 109

Usar las funciones `all` y `any` sobre un objeto `DataFrame`.

## Solución:

In [27]:
import pandas as pd

In [28]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [3, 2, 1, 0, -1], 'Z': [5, 7, 11, 13, 17]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  3   5
1  2  2   7
2  3  1  11
3  4  0  13
4  5 -1  17

In [29]:
df <= 3

X     Y      Z
0   True  True  False
1   True  True  False
2   True  True  False
3  False  True  False
4  False  True  False

In [30]:
(df <= 3).all()

X    False
Y     True
Z    False
dtype: bool

In [31]:
(df <= 3).any()

X     True
Y     True
Z    False
dtype: bool

# Ejercicio 110

Incrementar el número de columnas visibles de un `DataFrame` en **Jupyter**.

## Solución:

In [32]:
import pandas as pd

In [33]:
datos = {}

for i in range(1, 51):
    datos['Col%d' % i] = list(range(1, 11))

df = pd.DataFrame(data=datos)

df

Col1  Col2  Col3  Col4  Col5  Col6  Col7  Col8  Col9  Col10  ...  Col41  \
0     1     1     1     1     1     1     1     1     1      1  ...      1   
1     2     2     2     2     2     2     2     2     2      2  ...      2   
2     3     3     3     3     3     3     3     3     3      3  ...      3   
3     4     4     4     4     4     4     4     4     4      4  ...      4   
4     5     5     5     5     5     5     5     5     5      5  ...      5   
5     6     6     6     6     6     6     6     6     6      6  ...      6   
6     7     7     7     7     7     7     7     7     7      7  ...      7   
7     8     8     8     8     8     8     8     8     8      8  ...      8   
8     9     9     9     9     9     9     9     9     9      9  ...      9   
9    10    10    10    10    10    10    10    10    10     10  ...     10   

   Col42  Col43  Col44  Col45  Col46  Col47  Col48  Col49  Col50  
0      1      1      1      1      1      1      1      1      1  
1      2      2      2      2      2      2      2      2      2  
2      3      3      3      3      3      3      3      3      3  
3      4      4      4      4      4      4      4      4      4  
4      5      5      5      5      5      5      5      5      5  
5      6      6      6      6      6      6      6      6      6  
6      7      7      7      7      7      7      7      7      7  
7      8      8      8      8      8      8      8      8      8  
8      9      9      9      9      9      9      9      9      9  
9     10     10     10     10     10     10     10     10     10  

[10 rows x 50 columns]

In [34]:
pd.set_option('display.max_columns', 100)

In [35]:
df

Col1  Col2  Col3  Col4  Col5  Col6  Col7  Col8  Col9  Col10  Col11  Col12  \
0     1     1     1     1     1     1     1     1     1      1      1      1   
1     2     2     2     2     2     2     2     2     2      2      2      2   
2     3     3     3     3     3     3     3     3     3      3      3      3   
3     4     4     4     4     4     4     4     4     4      4      4      4   
4     5     5     5     5     5     5     5     5     5      5      5      5   
5     6     6     6     6     6     6     6     6     6      6      6      6   
6     7     7     7     7     7     7     7     7     7      7      7      7   
7     8     8     8     8     8     8     8     8     8      8      8      8   
8     9     9     9     9     9     9     9     9     9      9      9      9   
9    10    10    10    10    10    10    10    10    10     10     10     10   

   Col13  Col14  Col15  Col16  Col17  Col18  Col19  Col20  Col21  Col22  \
0      1      1      1      1      1      1      1      1      1      1   
1      2      2      2      2      2      2      2      2      2      2   
2      3      3      3      3      3      3      3      3      3      3   
3      4      4      4      4      4      4      4      4      4      4   
4      5      5      5      5      5      5      5      5      5      5   
5      6      6      6      6      6      6      6      6      6      6   
6      7      7      7      7      7      7      7      7      7      7   
7      8      8      8      8      8      8      8      8      8      8   
8      9      9      9      9      9      9      9      9      9      9   
9     10     10     10     10     10     10     10     10     10     10   

   Col23  Col24  Col25  Col26  Col27  Col28  Col29  Col30  Col31  Col32  \
0      1      1      1      1      1      1      1      1      1      1   
1      2      2      2      2      2      2      2      2      2      2   
2      3      3      3      3      3      3      3      3      3      3   
3      4      4      4      4      4      4      4      4      4      4   
4      5      5      5      5      5      5      5      5      5      5   
5      6      6      6      6      6      6      6      6      6      6   
6      7      7      7      7      7      7      7      7      7      7   
7      8      8      8      8      8      8      8      8      8      8   
8      9      9      9      9      9      9      9      9      9      9   
9     10     10     10     10     10     10     10     10     10     10   

   Col33  Col34  Col35  Col36  Col37  Col38  Col39  Col40  Col41  Col42  \
0      1      1      1      1      1      1      1      1      1      1   
1      2      2      2      2      2      2      2      2      2      2   
2      3      3      3      3      3      3      3      3      3      3   
3      4      4      4      4      4      4      4      4      4      4   
4      5      5      5      5      5      5      5      5      5      5   
5      6      6      6      6      6      6      6      6      6      6   
6      7      7      7      7      7      7      7      7      7      7   
7      8      8      8      8      8      8      8      8      8      8   
8      9      9      9      9      9      9      9      9      9      9   
9     10     10     10     10     10     10     10     10     10     10   

   Col43  Col44  Col45  Col46  Col47  Col48  Col49  Col50  
0      1      1      1      1      1      1      1      1  
1      2      2      2      2      2      2      2      2  
2      3      3      3      3      3      3      3      3  
3      4      4      4      4      4      4      4      4  
4      5      5      5      5      5      5      5      5  
5      6      6      6      6      6      6      6      6  
6      7      7      7      7      7      7      7      7  
7      8      8      8      8      8      8      8      8  
8      9      9      9      9      9      9      9      9  
9     10     10     10     10     10     10 

# Ejercicio 111

Obtener estadísticas básicas en un objeto `DataFrame` con la función `describe`.

## Solución:

In [36]:
import pandas as pd
import numpy as np

In [37]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [38]:
df.describe()

puntaje   intentos
count   8.000000  10.000000
mean   12.687500   1.900000
std     4.535082   0.875595
min     8.000000   1.000000
25%     8.000000   1.000000
50%    12.500000   2.000000
75%    16.125000   2.750000
max    19.000000   3.000000

# Ejercicio 112:

Incrementar el número de filas visibles de un `DataFrame` en **Jupyter**.

## Solución:

In [39]:
import pandas as pd

In [40]:
datos = {}

for i in range(1, 6):
    datos['Col%d' % i] = list(range(1, 101))

df = pd.DataFrame(data=datos)

df

Col1  Col2  Col3  Col4  Col5
0      1     1     1     1     1
1      2     2     2     2     2
2      3     3     3     3     3
3      4     4     4     4     4
4      5     5     5     5     5
5      6     6     6     6     6
6      7     7     7     7     7
7      8     8     8     8     8
8      9     9     9     9     9
9     10    10    10    10    10
10    11    11    11    11    11
11    12    12    12    12    12
12    13    13    13    13    13
13    14    14    14    14    14
14    15    15    15    15    15
15    16    16    16    16    16
16    17    17    17    17    17
17    18    18    18    18    18
18    19    19    19    19    19
19    20    20    20    20    20
20    21    21    21    21    21
21    22    22    22    22    22
22    23    23    23    23    23
23    24    24    24    24    24
24    25    25    25    25    25
25    26    26    26    26    26
26    27    27    27    27    27
27    28    28    28    28    28
28    29    29    29    29    29
29    30    30    30    30    30
..   ...   ...   ...   ...   ...
70    71    71    71    71    71
71    72    72    72    72    72
72    73    73    73    73    73
73    74    74    74    74    74
74    75    75    75    75    75
75    76    76    76    76    76
76    77    77    77    77    77
77    78    78    78    78    78
78    79    79    79    79    79
79    80    80    80    80    80
80    81    81    81    81    81
81    82    82    82    82    82
82    83    83    83    83    83
83    84    84    84    84    84
84    85    85    85    85    85
85    86    86    86    86    86
86    87    87    87    87    87
87    88    88    88    88    88
88    89    89    89    89    89
89    90    90    90    90    90
90    91    91    91    91    91
91    92    92    92    92    92
92    93    93    93    93    93
93    94    94    94    94    94
94    95    95    95    95    95
95    96    96    96    96    96
96    97    97    97    97    97
97    98    98    98    98    98
98    99    99    99    99    99
99   100   100   100   100   100

[100 rows x 5 columns]

In [41]:
pd.set_option('display.max_rows', 200)

In [42]:
df

Col1  Col2  Col3  Col4  Col5
0      1     1     1     1     1
1      2     2     2     2     2
2      3     3     3     3     3
3      4     4     4     4     4
4      5     5     5     5     5
5      6     6     6     6     6
6      7     7     7     7     7
7      8     8     8     8     8
8      9     9     9     9     9
9     10    10    10    10    10
10    11    11    11    11    11
11    12    12    12    12    12
12    13    13    13    13    13
13    14    14    14    14    14
14    15    15    15    15    15
15    16    16    16    16    16
16    17    17    17    17    17
17    18    18    18    18    18
18    19    19    19    19    19
19    20    20    20    20    20
20    21    21    21    21    21
21    22    22    22    22    22
22    23    23    23    23    23
23    24    24    24    24    24
24    25    25    25    25    25
25    26    26    26    26    26
26    27    27    27    27    27
27    28    28    28    28    28
28    29    29    29    29    29
29    30    30    30    30    30
30    31    31    31    31    31
31    32    32    32    32    32
32    33    33    33    33    33
33    34    34    34    34    34
34    35    35    35    35    35
35    36    36    36    36    36
36    37    37    37    37    37
37    38    38    38    38    38
38    39    39    39    39    39
39    40    40    40    40    40
40    41    41    41    41    41
41    42    42    42    42    42
42    43    43    43    43    43
43    44    44    44    44    44
44    45    45    45    45    45
45    46    46    46    46    46
46    47    47    47    47    47
47    48    48    48    48    48
48    49    49    49    49    49
49    50    50    50    50    50
50    51    51    51    51    51
51    52    52    52    52    52
52    53    53    53    53    53
53    54    54    54    54    54
54    55    55    55    55    55
55    56    56    56    56    56
56    57    57    57    57    57
57    58    58    58    58    58
58    59    59    59    59    59
59    60    60    60    60    60
60    61    61    61    61    61
61    62    62    62    62    62
62    63    63    63    63    63
63    64    64    64    64    64
64    65    65    65    65    65
65    66    66    66    66    66
66    67    67    67    67    67
67    68    68    68    68    68
68    69    69    69    69    69
69    70    70    70    70    70
70    71    71    71    71    71
71    72    72    72    72    72
72    73    73    73    73    73
73    74    74    74    74    74
74    75    75    75    75    75
75    76    76    76    76    76
76    77    77    77    77    77
77    78    78    78    78    78
78    79    79    79    79    79
79    80    80    80    80    80
80    81    81    81    81    81
81    82    82    82    82    82
82    83    83    83    83    83
83    84    84    84    84    84
84    85    85    85    85    85
85    86    86    86    86    86
86    87    87    87    87    87
87    88    88    88    88    88
88    89    89    89    89    89
89    90    90    90    90    90
90    91    91    91    91    91
91    92    92    92    92    92
92    93    93    93    93    93
93    94    94    94    94    94
94    95    95    95    95    95
95    96    96    96    96    96
96    97    97    97    97    97
97    98    98    98    98    98
98    99    99    99    99    99
99   100   100   100   100   100

# Ejercicio 113

Obtener registros a partir de sus índices numéricos con `iloc`.

## Solución:

In [43]:
import pandas as pd

In [44]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [45]:
resultado = df.iloc[[1, 3]]

resultado

X  Y  Z
1  2  4  3
3  4  2  7

# Ejercicio 114

Calcular el elemento mayor por cada fila y columna con la función `max`.

## Solución:

In [46]:
import pandas as pd

In [47]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [48]:
df.max(axis=1)

0     5
1     4
2     5
3     7
4    11
dtype: int64

In [49]:
df.max()

X     5
Y     5
Z    11
dtype: int64

# Ejercicio 115

Reemplazar los NaN (nulo) con un valor específico usando la función `fillna`.

## Solución:

In [50]:
import pandas as pd
import numpy as np

In [51]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [52]:
resultado = df.fillna(0)

resultado

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      0.0         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      0.0         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

# Ejercicio 116

Calcular el promedio y la mediana por cada fila y columna con `mean` y `median`.

## Solución:

In [53]:
import pandas as pd

In [54]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [55]:
df.median(axis=1)

0    2.0
1    3.0
2    3.0
3    4.0
4    5.0
dtype: float64

In [56]:
df.mean(axis=1)

0    2.666667
1    3.000000
2    3.666667
3    4.333333
4    5.666667
dtype: float64

In [57]:
# Mediana por cada columna:
df.median(axis=0)

X    3.0
Y    3.0
Z    5.0
dtype: float64

In [58]:
# Promedio por cada columna:
df.mean(axis=0)

X    3.0
Y    3.0
Z    5.6
dtype: float64

# Ejercicio 117

Calcular el elemento menor por cada fila y columna con la función `min`.

## Solución:

In [59]:
import pandas as pd

In [60]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 0], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  0  11

In [61]:
df.min(axis=1)

0    1
1    2
2    3
3    2
4    0
dtype: int64

In [62]:
df.min(axis=0)

X    1
Y    0
Z    2
dtype: int64

# Ejercicio 118

Convertir los índices de un `DataFrame` en una columna con la función `reset_index`.

## Solución:

In [63]:
import pandas as pd

In [64]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 0], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  0  11

In [65]:
df.reset_index(level=0, inplace=True)

df

index  X  Y   Z
0     a  1  5   2
1     b  2  4   3
2     c  3  3   5
3     d  4  2   7
4     e  5  0  11

In [66]:
print(df.to_string(index=False))

index  X  Y   Z
    a  1  5   2
    b  2  4   3
    c  3  3   5
    d  4  2   7
    e  5  0  11


# Ejercicio 119

Seleccionar registros que cumplan con una hora específica con `at_time`.

## Solución:

In [67]:
import pandas as pd

In [68]:
fechas = pd.date_range('2023-08-16', periods=8, freq='6H')
datos = {'A': list(range(1, 9))}

df = pd.DataFrame(data=datos, index=fechas)

df

A
2023-08-16 00:00:00  1
2023-08-16 06:00:00  2
2023-08-16 12:00:00  3
2023-08-16 18:00:00  4
2023-08-17 00:00:00  5
2023-08-17 06:00:00  6
2023-08-17 12:00:00  7
2023-08-17 18:00:00  8

In [69]:
df.at_time('06:00')

A
2023-08-16 06:00:00  2
2023-08-17 06:00:00  6

# Ejercicio 120

Usar la función `set_value` para cambiar un elemento específico en un `DataFrame`.

## Solución:

In [70]:
import pandas as pd
from string import ascii_lowercase

In [71]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 0], 'Z': [2, 3, 5, 7, 11]}

indices = list(ascii_lowercase[:5])

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  0  11

In [72]:
df.set_value('a', 'X', 0)

C:\Users\johno\Anaconda3-2018.12\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  """Entry point for launching an IPython kernel.


X  Y   Z
a  0  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  0  11

In [73]:
df.iat[0, 0] = 0

df

X  Y   Z
a  0  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  0  11

In [74]:
df.at['e', 'Z'] = 29

df

X  Y   Z
a  0  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  0  29

# Ejercicio 121

Calcular la productoria por cada fila y columna de un `DataFrame`.

$\prod_{i=1}^{n}k_{i}=k_1\times k_2 \times ...\times k_n$

## Solución:

In [75]:
import pandas as pd

In [76]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 0], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  0  11

In [77]:
df.product()

X     120
Y       0
Z    2310
dtype: int64

In [78]:
df.product(axis=1)

0    10
1    24
2    45
3    56
4     0
dtype: int64

In [79]:
df.prod(axis=1)

0    10
1    24
2    45
3    56
4     0
dtype: int64

# Ejercicio 122

Contar los elementos que son nulos (NaN) en un objeto `DataFrame`.

## Solución:

In [80]:
import pandas as pd
import numpy as np

In [81]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [82]:
df.isnull()

nombre  puntaje  intentos  califico
a   False    False     False     False
b   False    False     False     False
c   False    False     False     False
d   False     True     False     False
e   False    False     False     False
f   False    False     False     False
g   False    False     False     False
h   False     True     False     False
i   False    False     False     False
j   False    False     False     False

In [83]:
df.isnull().values

array([[False, False, False, False],
       [False, False, False, False],
       [False, False, False, False],
       [False,  True, False, False],
       [False, False, False, False],
       [False, False, False, False],
       [False, False, False, False],
       [False,  True, False, False],
       [False, False, False, False],
       [False, False, False, False]])

In [84]:
type(df.isnull().values)

numpy.ndarray

In [85]:
df.isnull().values.sum()

2

# Ejercicio 123

Redondear valores reales con la función `DataFrame.round`.

## Solución:

In [86]:
import pandas as pd

In [87]:
numeros = [[3.1415, 2.717], [1.41, 1]]

df = pd.DataFrame(numeros)

df

0      1
0  3.1415  2.717
1  1.4100  1.000

In [88]:
df.round()

0    1
0  3.0  3.0
1  1.0  1.0

In [89]:
df.round(5)

0      1
0  3.1415  2.717
1  1.4100  1.000

In [90]:
df.round(8)

0      1
0  3.1415  2.717
1  1.4100  1.000

In [91]:
df.round(4)

0      1
0  3.1415  2.717
1  1.4100  1.000

# Ejercicio 124

Eliminar filas o registros con la función `DataFrame.drop`.

## Solución:

In [92]:
import pandas as pd
import numpy as np

In [93]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [94]:
df.drop(['f', 'h'])

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
g    Julio     13.5         1       Sí
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [95]:
df.drop(df.index[[5, 7]])

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
g    Julio     13.5         1       Sí
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

# Ejercicio 125

Encontrar la cantidad de elementos únicos por cada fila y columna con `nunique`.

## Solución:

In [96]:
import pandas as pd

In [97]:
datos = {'X': [1, 2, 3, 4, 1], 'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

df = pd.DataFrame(data=datos)

df

X  Y  Z
0  1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [98]:
df.nunique(axis=1)

0    3
1    1
2    2
3    3
4    2
dtype: int64

In [99]:
df.nunique(axis=0)

X    4
Y    4
Z    4
dtype: int64

# Ejercicio 126

Reiniciar un índice numérico después de eliminar algunos registros con `reset_index`.

## Solución:

In [100]:
import pandas as pd
import numpy as np

In [101]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

estudiantes = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=estudiantes)

df

nombre  puntaje  intentos califico
0    Oliva     11.5         1       Sí
1  Daniela      8.0         3       No
2     Juan     15.5         2       Sí
3   Germán      NaN         3       No
4   Edward      8.0         2       No
5     Alex     19.0         3       Sí
6    Julio     13.5         1       Sí
7    Edgar      NaN         1       No
8    Angie      8.0         2       No
9   Irlesa     18.0         1       Sí

In [102]:
df = df.drop([5, 7])

df

nombre  puntaje  intentos califico
0    Oliva     11.5         1       Sí
1  Daniela      8.0         3       No
2     Juan     15.5         2       Sí
3   Germán      NaN         3       No
4   Edward      8.0         2       No
6    Julio     13.5         1       Sí
8    Angie      8.0         2       No
9   Irlesa     18.0         1       Sí

In [103]:
df = df.reset_index()

df

index   nombre  puntaje  intentos califico
0      0    Oliva     11.5         1       Sí
1      1  Daniela      8.0         3       No
2      2     Juan     15.5         2       Sí
3      3   Germán      NaN         3       No
4      4   Edward      8.0         2       No
5      6    Julio     13.5         1       Sí
6      8    Angie      8.0         2       No
7      9   Irlesa     18.0         1       Sí

In [104]:
df.drop(['index'], axis=1, inplace=True)

df

nombre  puntaje  intentos califico
0    Oliva     11.5         1       Sí
1  Daniela      8.0         3       No
2     Juan     15.5         2       Sí
3   Germán      NaN         3       No
4   Edward      8.0         2       No
5    Julio     13.5         1       Sí
6    Angie      8.0         2       No
7   Irlesa     18.0         1       Sí

# Ejercicio 127

Agregar un prefijo y postfijo a las columnas de un objeto `DataFrame`.

## Solución:

In [105]:
import pandas as pd

In [106]:
datos = {1: [1, 2, 3, 4, 5], 2: [5, 4, 3, 2, 0], 3: [2, 3, 5, 7, 11]}

df = pd.DataFrame(datos)
df

1  2   3
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  0  11

In [107]:
df.add_prefix('Columna')

Columna1  Columna2  Columna3
0         1         5         2
1         2         4         3
2         3         3         5
3         4         2         7
4         5         0        11

In [108]:
df.add_suffix('_col')

1_col  2_col  3_col
0      1      5      2
1      2      4      3
2      3      3      5
3      4      2      7
4      5      0     11

# Ejercicio 128

Seleccionar un conjunto de registros a partir de un rango horario.

## Solución:

In [109]:
import pandas as pd

In [110]:
indices = pd.date_range('2018-04-09', periods=4, freq='1D30min')

df = pd.DataFrame({'A': [100, 200, 300, 400]}, index=indices)

df

A
2018-04-09 00:00:00  100
2018-04-10 00:30:00  200
2018-04-11 01:00:00  300
2018-04-12 01:30:00  400

In [111]:
df.between_time('00:30', '01:00')

A
2018-04-10 00:30:00  200
2018-04-11 01:00:00  300

# Ejercicio 129

Combinar dos objetos `DataFrame` con la función `concat`.

## Solución:

In [112]:
import pandas as pd

In [113]:
programadores = ['Edward', 'Germán', 'Daniela', 'John']

df_1 = pd.DataFrame({'programador': programadores})

df_1

programador
0      Edward
1      Germán
2     Daniela
3        John

In [114]:
lenguajes = ['C++', 'Java', 'Scala', 'Python']

df_2 = pd.DataFrame({'lenguaje': lenguajes})

df_2

lenguaje
0      C++
1     Java
2    Scala
3   Python

In [115]:
df = pd.concat([df_1, df_2], axis=1)

df

programador lenguaje
0      Edward      C++
1      Germán     Java
2     Daniela    Scala
3        John   Python

# Ejercicio 130

Eliminar registros duplicados de un `DataFrame` a partir de una columna.

## Solución:

In [116]:
import pandas as pd

In [117]:
programadores = ['Edward', 'Germán', 'Daniela', 'John', 'Germán', 'John']
lenguajes = ['C++', 'Java', 'Scala', 'Python', 'PHP', 'JavaScript']

datos = {'programador': programadores, 'lenguaje': lenguajes}

df = pd.DataFrame(datos)

df

programador    lenguaje
0      Edward         C++
1      Germán        Java
2     Daniela       Scala
3        John      Python
4      Germán         PHP
5        John  JavaScript

In [118]:
df.drop_duplicates('programador')

programador lenguaje
0      Edward      C++
1      Germán     Java
2     Daniela    Scala
3        John   Python

In [119]:
df.drop_duplicates('programador', keep='last')

programador    lenguaje
0      Edward         C++
2     Daniela       Scala
4      Germán         PHP
5        John  JavaScript

# Ejercicio 131

Cambiar el orden de las filas de forma aleatoria con la función `sample`.

## Solución:

In [120]:
import pandas as pd
import numpy as np

In [121]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

estudiantes = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=estudiantes)

df

nombre  puntaje  intentos califico
0    Oliva     11.5         1       Sí
1  Daniela      8.0         3       No
2     Juan     15.5         2       Sí
3   Germán      NaN         3       No
4   Edward      8.0         2       No
5     Alex     19.0         3       Sí
6    Julio     13.5         1       Sí
7    Edgar      NaN         1       No
8    Angie      8.0         2       No
9   Irlesa     18.0         1       Sí

In [122]:
df.sample(frac=1)

nombre  puntaje  intentos califico
5     Alex     19.0         3       Sí
8    Angie      8.0         2       No
1  Daniela      8.0         3       No
7    Edgar      NaN         1       No
4   Edward      8.0         2       No
2     Juan     15.5         2       Sí
3   Germán      NaN         3       No
6    Julio     13.5         1       Sí
0    Oliva     11.5         1       Sí
9   Irlesa     18.0         1       Sí

# Ejercicio 132

Verificar con `equals` si dos objetos `DataFrame` contienen los mismos elementos.

## Solución:

In [123]:
import pandas as pd

In [124]:
datos = {'X': [1, 2, 3, 4, 1], 'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

df = pd.DataFrame(data=datos)

df

X  Y  Z
0  1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [125]:
igual = pd.DataFrame(data=datos)

igual

X  Y  Z
0  1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [126]:
df.equals(igual)

True

In [127]:
datos['X'][0] = -1

datos

{'X': [-1, 2, 3, 4, 1], 'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

In [128]:
no_igual = pd.DataFrame(data=datos)

no_igual

X  Y  Z
0 -1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [129]:
df.equals(no_igual)

False

# Ejercicio 133

Convertir una columna de tipo cadena de caracteres a fecha con `to_datetime`.

## Solución:

In [130]:
import pandas as pd

In [131]:
fechas = ['2018/12/31', '2019/01/31', '2023/05/21']

serie = pd.Series(fechas)

serie

0    2018/12/31
1    2019/01/31
2    2023/05/21
dtype: object

In [132]:
serie = pd.to_datetime(pd.Series(serie))

serie

0   2018-12-31
1   2019-01-31
2   2023-05-21
dtype: datetime64[ns]

In [133]:
df = pd.DataFrame(serie)

df

0
0 2018-12-31
1 2019-01-31
2 2023-05-21

In [134]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 1 columns):
0    3 non-null datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 104.0 bytes


# Ejercicio 134

Usar `first` y `last` para obtener los primeros y últimos registros para una cantidad de días arbitraria

## Solución:

In [135]:
import pandas as pd

In [136]:
indices = pd.date_range('2023-01-01', periods=8, freq='2D')

indices

DatetimeIndex(['2023-01-01', '2023-01-03', '2023-01-05', '2023-01-07',
               '2023-01-09', '2023-01-11', '2023-01-13', '2023-01-15'],
              dtype='datetime64[ns]', freq='2D')

In [137]:
contenido = list(range(1, 9))
datos = {'A': contenido}

df = pd.DataFrame(data=datos, index=indices)

df

A
2023-01-01  1
2023-01-03  2
2023-01-05  3
2023-01-07  4
2023-01-09  5
2023-01-11  6
2023-01-13  7
2023-01-15  8

In [138]:
df.first('3D')

A
2023-01-01  1
2023-01-03  2

In [139]:
df.first('5D')

A
2023-01-01  1
2023-01-03  2
2023-01-05  3

In [140]:
df.last('3D')

A
2023-01-13  7
2023-01-15  8

In [141]:
df.last('5D')

A
2023-01-11  6
2023-01-13  7
2023-01-15  8

# Ejercicio 135

Renombrar una columna de un `DataFrame` con la función `rename`.

## Solución:

In [142]:
import pandas as pd

In [143]:
datos = {'Columna1': [1, 2, 3], 'col2': [4, 5, 6], 'Col3': [7, 8, 9]}

df = pd.DataFrame(data=datos)

df

Columna1  col2  Col3
0         1     4     7
1         2     5     8
2         3     6     9

In [144]:
df.rename(columns={'Columna1': 'col1'})

col1  col2  Col3
0     1     4     7
1     2     5     8
2     3     6     9

# Ejercicio 136

Obtener los datos de una columna como una lista Python con `tolist`.

## Solución:

In [145]:
import pandas as pd

In [146]:
datos = {'X': [1, 2, 3, 4, 1], 'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

df = pd.DataFrame(data=datos)

df

X  Y  Z
0  1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [147]:
lista_valores = df['Y'].tolist()

type(lista_valores)

list

In [148]:
lista_valores

[5, 2, 3, 2, 0]

# Ejercicio 137

Seleccionar filas de un `DataFrame` a partir su posición con `take`.

## Solución:

In [149]:
import pandas as pd

In [150]:
datos = {'X': [1, 2, 3, 4, 1], 'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

indices = [0, 2, 1, 4, 3]

df = pd.DataFrame(data=datos, index=indices)

df

X  Y  Z
0  1  5  2
2  2  2  2
1  3  3  5
4  4  2  7
3  1  0  1

In [151]:
df.take([0, 3])

X  Y  Z
0  1  5  2
4  4  2  7

# Ejercicio 138

Seleccionar una columna de un `DataFrame` como índice con `set_index`.

## Solución:

In [152]:
import pandas as pd

In [153]:
documentos = ['1951', '8546', '6547', '6541', '9537']

datos = {'X': documentos, 'Y': [5, 2, 3, 2, 2], 'Z': [2, 2, 8, 7, 7]}

df = pd.DataFrame(data=datos)

df

X  Y  Z
0  1951  5  2
1  8546  2  2
2  6547  3  8
3  6541  2  7
4  9537  2  7

In [154]:
df = df.set_index('X')

df

Y  Z
X         
1951  5  2
8546  2  2
6547  3  8
6541  2  7
9537  2  7

In [155]:
df.loc[['6547', '8546']]

Y  Z
X         
6547  3  8
8546  2  2

# Ejercicio 139

Truncar un `DataFrame` por filas y columnas con la función `truncate`.

## Solución:

In [156]:
import pandas as pd

In [157]:
datos = {'X': [1, 2, 3, 4, 1], 'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

df = pd.DataFrame(data=datos)

df

X  Y  Z
0  1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [158]:
df.truncate(before=2, after=4)

X  Y  Z
2  3  3  5
3  4  2  7
4  1  0  1

In [159]:
df.truncate(before='Y', after='Z', axis=1)

Y  Z
0  5  2
1  2  2
2  3  5
3  2  7
4  0  1

In [160]:
df['X']

0    1
1    2
2    3
3    4
4    1
Name: X, dtype: int64

In [161]:
df['X'].truncate(before=2, after=3)

2    3
3    4
Name: X, dtype: int64

 # Ejercicio 140
 
 Crear un `DataFrame` a partir de arreglos `NumPy` y crear columnas e índices.
 
 ## Solución

In [162]:
import pandas as pd
import numpy as np

In [163]:
dtype = [('Columna1', 'int32'), ('Columna2', 'float32'), ('Columna3', 'float64')]

arreglo = np.zeros(15, dtype=dtype)

arreglo

array([(0, 0., 0.), (0, 0., 0.), (0, 0., 0.), (0, 0., 0.), (0, 0., 0.),
       (0, 0., 0.), (0, 0., 0.), (0, 0., 0.), (0, 0., 0.), (0, 0., 0.),
       (0, 0., 0.), (0, 0., 0.), (0, 0., 0.), (0, 0., 0.), (0, 0., 0.)],
      dtype=[('Columna1', '<i4'), ('Columna2', '<f4'), ('Columna3', '<f8')])

In [164]:
indices = ['Indice{}'.format(i) for i in range(1, len(arreglo) + 1)]

indices

['Indice1',
 'Indice2',
 'Indice3',
 'Indice4',
 'Indice5',
 'Indice6',
 'Indice7',
 'Indice8',
 'Indice9',
 'Indice10',
 'Indice11',
 'Indice12',
 'Indice13',
 'Indice14',
 'Indice15']

In [165]:
df = pd.DataFrame(data=arreglo, index=indices)

df

Columna1  Columna2  Columna3
Indice1          0       0.0       0.0
Indice2          0       0.0       0.0
Indice3          0       0.0       0.0
Indice4          0       0.0       0.0
Indice5          0       0.0       0.0
Indice6          0       0.0       0.0
Indice7          0       0.0       0.0
Indice8          0       0.0       0.0
Indice9          0       0.0       0.0
Indice10         0       0.0       0.0
Indice11         0       0.0       0.0
Indice12         0       0.0       0.0
Indice13         0       0.0       0.0
Indice14         0       0.0       0.0
Indice15         0       0.0       0.0

# Ejercicio 141

Agregar una nueva columna a un `DataFrame` por medio de la función `assign`.

## Solución:

In [166]:
import pandas as pd

In [167]:
temperatura = {'C': [13.0, 23.0]}
indices = ['Isnos', 'Pitalito']

df = pd.DataFrame(data=temperatura, index=indices)

df

C
Isnos     13.0
Pitalito  23.0

In [168]:
df.assign(F=lambda x: x.C * 9 / 5 + 32)

C     F
Isnos     13.0  55.4
Pitalito  23.0  73.4

# Ejercicio 142

Seleccionar el índice de una columna de un `DataFrame` con el elemento mayor.

## Solución:

In [169]:
import pandas as pd

In [170]:
datos = {'X': [1, 2, 3, 4, 1], 'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

df = pd.DataFrame(data=datos)

df

X  Y  Z
0  1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [171]:
df['X'].argmax()

C:\Users\johno\Anaconda3-2018.12\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: 
The current behaviour of 'Series.argmax' is deprecated, use 'idxmax'
instead.
The behavior of 'argmax' will be corrected to return the positional
maximum in the future. For now, use 'series.values.argmax' or
'np.argmax(np.array(values))' to get the position of the maximum
row.
  """Entry point for launching an IPython kernel.


3

In [172]:
df['Y'].argmax()

C:\Users\johno\Anaconda3-2018.12\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: 
The current behaviour of 'Series.argmax' is deprecated, use 'idxmax'
instead.
The behavior of 'argmax' will be corrected to return the positional
maximum in the future. For now, use 'series.values.argmax' or
'np.argmax(np.array(values))' to get the position of the maximum
row.
  """Entry point for launching an IPython kernel.


0

In [173]:
df['Z'].argmax()

C:\Users\johno\Anaconda3-2018.12\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: 
The current behaviour of 'Series.argmax' is deprecated, use 'idxmax'
instead.
The behavior of 'argmax' will be corrected to return the positional
maximum in the future. For now, use 'series.values.argmax' or
'np.argmax(np.array(values))' to get the position of the maximum
row.
  """Entry point for launching an IPython kernel.


3

In [174]:
df['X'].idxmax()

3

# Ejercicio 143

Comprobar si una columna existe en un `DataFrame` con el operador Python `in`.

## Solución:

In [175]:
import pandas as pd

In [176]:
datos = {'X': [1, 2, 3, 4, 1], 'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

df = pd.DataFrame(data=datos)

df

X  Y  Z
0  1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [177]:
'X' in df.columns

True

In [178]:
'x' in df.columns

False

In [179]:
'W' in df.columns

False

# Ejercicio 144

Calcular la transpuesta de un objeto `DataFrame` con `transpose` o `T`.

## Solución:

In [180]:
import pandas as pd

In [181]:
datos = {'X': [1, 2, 3, 4, 1], 'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

df = pd.DataFrame(data=datos)

df

X  Y  Z
0  1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [182]:
df.transpose()

0  1  2  3  4
X  1  2  3  4  1
Y  5  2  3  2  0
Z  2  2  5  7  1

In [183]:
df.T

0  1  2  3  4
X  1  2  3  4  1
Y  5  2  3  2  0
Z  2  2  5  7  1

# Ejercicio 145

Reemplazar valores en un `DataFrame` con la función `replace`.

## Solución:

In [184]:
import pandas as pd

In [185]:
datos = {'X': [1, 2, 3, 4, 1], 'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

df = pd.DataFrame(data=datos)

df

X  Y  Z
0  1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [186]:
df.replace(2, 0)

X  Y  Z
0  1  5  0
1  0  0  0
2  3  3  5
3  4  0  7
4  1  0  1

In [187]:
df

X  Y  Z
0  1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [188]:
df.replace([5, 2, 3], 0)

X  Y  Z
0  1  0  0
1  0  0  0
2  0  0  0
3  4  0  7
4  1  0  1

In [189]:
df

X  Y  Z
0  1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [190]:
df.replace([1, 5], [-1, -5])

X  Y  Z
0 -1 -5  2
1  2  2  2
2  3  3 -5
3  4  2  7
4 -1  0 -1

# Ejercicio 146

Obtener filas de un `DataFrame` con `iloc`.

## Solución:

In [191]:
import pandas as pd

In [192]:
datos = {'X': [1, 2, 3, 4, 1], 'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

df = pd.DataFrame(data=datos)

df

X  Y  Z
0  1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [193]:
df.iloc[0]

X    1
Y    5
Z    2
Name: 0, dtype: int64

In [194]:
type(df.iloc[0])

pandas.core.series.Series

In [195]:
df.iloc[4]

X    1
Y    0
Z    1
Name: 4, dtype: int64

In [196]:
# df.iloc[10]

# Ejercicio 147

Obtener los tipos de datos de las columnas de un `DataFrame` con `dtypes`.

## Solución:

In [197]:
import pandas as pd
import numpy as np

In [198]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [199]:
df.dtypes

nombre       object
puntaje     float64
intentos      int64
califico     object
dtype: object

In [200]:
type(df.dtypes)

pandas.core.series.Series

# Ejercicio 148

Ordenar los valores de las columnas de un `DataFrame` con `sort_values`.

## Solución:

In [201]:
import pandas as pd
import numpy as np

In [202]:
datos = {'X': ['D', 'C', 'A', 'D', np.nan], 'Y': [5, 2, 3, 2, 0], 'Z': [3, 3, 5, 7, 1]}

df = pd.DataFrame(data=datos)

df

X  Y  Z
0    D  5  3
1    C  2  3
2    A  3  5
3    D  2  7
4  NaN  0  1

In [203]:
df.sort_values(by=['X'])

X  Y  Z
2    A  3  5
1    C  2  3
0    D  5  3
3    D  2  7
4  NaN  0  1

In [204]:
df.sort_values(by=['X', 'Y'])

X  Y  Z
2    A  3  5
1    C  2  3
3    D  2  7
0    D  5  3
4  NaN  0  1

In [205]:
df.sort_values(by=['X'], ascending=False)

X  Y  Z
0    D  5  3
3    D  2  7
1    C  2  3
2    A  3  5
4  NaN  0  1

In [206]:
df.sort_values(by=['X'], ascending=False, na_position='first')

X  Y  Z
4  NaN  0  1
0    D  5  3
3    D  2  7
1    C  2  3
2    A  3  5

# Ejercicio 149

Agregar datos a un `DataFrame` vacío con la función `append`.

## Solución:

In [207]:
import pandas as pd

In [208]:
df = pd.DataFrame()

df

Empty DataFrame
Columns: []
Index: []

In [209]:
df.size

0

In [210]:
datos = {'X': [1, 2, 3, 4, 1], 'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

df = df.append(pd.DataFrame(data=datos))

df

X  Y  Z
0  1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [211]:
df.size

15

# Ejercicio 150

Ordenar los índices de un `DataFrame` con la función `sort_index`.

## Solución:

In [212]:
import pandas as pd

In [213]:
datos = {'X': [1, 2, 3, 4, 1], 'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

indices = [2, 3, 4, 0, 1]

df = pd.DataFrame(data=datos, index=indices)

df

X  Y  Z
2  1  5  2
3  2  2  2
4  3  3  5
0  4  2  7
1  1  0  1

In [214]:
df = df.sort_index()

df

X  Y  Z
0  4  2  7
1  1  0  1
2  1  5  2
3  2  2  2
4  3  3  5

In [215]:
df.index = ['d', 'e', 'a', 'c', 'b']

df

X  Y  Z
d  4  2  7
e  1  0  1
a  1  5  2
c  2  2  2
b  3  3  5

In [216]:
df.sort_index()

X  Y  Z
a  1  5  2
b  3  3  5
c  2  2  2
d  4  2  7
e  1  0  1

# Ejercicio 151

Ordenar los registros de un `DataFrame` por múltiples columnas.

## Solución:

In [217]:
import pandas as pd
import numpy as np

In [218]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [219]:
df.sort_values(['puntaje', 'califico'])

nombre  puntaje  intentos califico
b  Daniela      8.0         3       No
e   Edward      8.0         2       No
i    Angie      8.0         2       No
a    Oliva     11.5         1       Sí
g    Julio     13.5         1       Sí
c     Juan     15.5         2       Sí
j   Irlesa     18.0         1       Sí
f     Alex     19.0         3       Sí
d   Germán      NaN         3       No
h    Edgar      NaN         1       No

In [220]:
df.sort_values(['puntaje', 'califico'], ascending=[False, True])

nombre  puntaje  intentos califico
f     Alex     19.0         3       Sí
j   Irlesa     18.0         1       Sí
c     Juan     15.5         2       Sí
g    Julio     13.5         1       Sí
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
e   Edward      8.0         2       No
i    Angie      8.0         2       No
d   Germán      NaN         3       No
h    Edgar      NaN         1       No

# Ejercicio 152

Obtener los `n` registros menores de un `DataFrame` con `nsmallest`.

## Solución:

In [221]:
import pandas as pd
import numpy as np

In [222]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [223]:
df.nsmallest(3, 'intentos')

nombre  puntaje  intentos califico
a  Oliva     11.5         1       Sí
g  Julio     13.5         1       Sí
h  Edgar      NaN         1       No

In [224]:
df.nsmallest(4, 'intentos')

nombre  puntaje  intentos califico
a   Oliva     11.5         1       Sí
g   Julio     13.5         1       Sí
h   Edgar      NaN         1       No
j  Irlesa     18.0         1       Sí

In [225]:
df.nsmallest(3, 'intentos', keep='last')

nombre  puntaje  intentos califico
j  Irlesa     18.0         1       Sí
h   Edgar      NaN         1       No
g   Julio     13.5         1       Sí

In [226]:
df.nsmallest(5, 'intentos')

nombre  puntaje  intentos califico
a   Oliva     11.5         1       Sí
g   Julio     13.5         1       Sí
h   Edgar      NaN         1       No
j  Irlesa     18.0         1       Sí
c    Juan     15.5         2       Sí

In [227]:
df.sort_values('intentos', ascending=True).head(3)

nombre  puntaje  intentos califico
a  Oliva     11.5         1       Sí
g  Julio     13.5         1       Sí
h  Edgar      NaN         1       No

# Ejercicio 153

Obtener los `n` registros mayores de un `DataFrame` con `nlargest`.

## Solución:

In [228]:
import pandas as pd
import numpy as np

In [229]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [230]:
df.nlargest(3, 'intentos')

nombre  puntaje  intentos califico
b  Daniela      8.0         3       No
d   Germán      NaN         3       No
f     Alex     19.0         3       Sí

In [231]:
df.nlargest(2, 'intentos')

nombre  puntaje  intentos califico
b  Daniela      8.0         3       No
d   Germán      NaN         3       No

In [232]:
df.nlargest(2, 'intentos', keep='last')

nombre  puntaje  intentos califico
f    Alex     19.0         3       Sí
d  Germán      NaN         3       No

In [233]:
df.nlargest(3, ['intentos', 'puntaje'])

nombre  puntaje  intentos califico
f     Alex     19.0         3       Sí
b  Daniela      8.0         3       No
d   Germán      NaN         3       No

# Ejercicio 154

Cambiar el tipo de una columna de `float` a `int` con la función `astype`.

## Solución:

In [234]:
import pandas as pd
import numpy as np

In [235]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [236]:
df.dtypes

nombre       object
puntaje     float64
intentos      int64
califico     object
dtype: object

In [237]:
df.puntaje = df.puntaje.fillna(0)

df.puntaje = df.puntaje.astype(int)

In [238]:
df

nombre  puntaje  intentos califico
a    Oliva       11         1       Sí
b  Daniela        8         3       No
c     Juan       15         2       Sí
d   Germán        0         3       No
e   Edward        8         2       No
f     Alex       19         3       Sí
g    Julio       13         1       Sí
h    Edgar        0         1       No
i    Angie        8         2       No
j   Irlesa       18         1       Sí

In [239]:
df.dtypes

nombre      object
puntaje      int32
intentos     int64
califico    object
dtype: object

# Ejercicio 155

Remover valores infinitos (`-np.inf` y `np.inf`) de un `DataFrame` con `replace`.

## Solución:

In [240]:
import numpy as np
import pandas as pd

In [241]:
df = pd.DataFrame([2, 3, 5, 7, np.inf, 11, -np.inf])

df

0
0   2.0
1   3.0
2   5.0
3   7.0
4   inf
5  11.0
6  -inf

In [242]:
df = df.replace([np.inf, -np.inf], np.nan)

df

0
0   2.0
1   3.0
2   5.0
3   7.0
4   NaN
5  11.0
6   NaN

# Ejercicio 156

Agregar una columna en una posición específica con la función `DataFrame.insert`.

## Solución:

In [243]:
import pandas as pd

In [244]:
datos = {'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

df = pd.DataFrame(data=datos)

df

Y  Z
0  5  2
1  2  2
2  3  5
3  2  7
4  0  1

In [245]:
valores = [7, 8, 9, 2, 3]

posicion = 0

df.insert(loc=posicion, column='X', value=valores)

df

X  Y  Z
0  7  5  2
1  8  2  2
2  9  3  5
3  2  2  7
4  3  0  1

# Ejercicio 157

Crear un `DataFrame` a partir de una matriz (lista de listas) Python.

## Solución

In [246]:
import pandas as pd

In [247]:
matriz = [['X', 'Y'], [3, 2], [7, 3]]

matriz

[['X', 'Y'], [3, 2], [7, 3]]

In [248]:
encabezado = matriz.pop(0)

encabezado

['X', 'Y']

In [249]:
df = pd.DataFrame(data=matriz, columns=encabezado)

df

X  Y
0  3  2
1  7  3

# Ejercicio 158

Calcular el número de tipos de datos por cada categoría con `get_dtype_counts`.

## Solución:

In [250]:
import pandas as pd
import numpy as np

In [251]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [252]:
df.get_dtype_counts()

float64    1
int64      1
object     2
dtype: int64

# Ejercicio 159

Obtener la posición numérica de una columna a partir de su nombre con `get_loc`.

## Solución:

In [253]:
import pandas as pd

In [254]:
datos = {'X': [1, 2, 3, 4, 1], 'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

df = pd.DataFrame(data=datos)

df

X  Y  Z
0  1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [255]:
df.columns

Index(['X', 'Y', 'Z'], dtype='object')

In [256]:
df.columns.get_loc('Y')

1

# Ejercicio 160

Contar el número de columnas que tiene un `DataFrame`.

## Solución:

In [257]:
import pandas as pd

In [258]:
datos = {'X': [1, 2, 3, 4, 1], 'Y': [5, 2, 3, 2, 0], 'Z': [2, 2, 5, 7, 1]}

df = pd.DataFrame(data=datos)

df

X  Y  Z
0  1  5  2
1  2  2  2
2  3  3  5
3  4  2  7
4  1  0  1

In [259]:
df.columns

Index(['X', 'Y', 'Z'], dtype='object')

In [260]:
len(df.columns)

3